### ISSUE TITLE
When user changed to white warpgates don't change

### ISSUE THREAD

**Issue by Skilly146** [anicolao/npa#1]
when I use the 'w' hotkey it correctly changes all of my stars to white but the warpgate highlight effect is still the original color. Ex: if my stars are blue with one having a warpgate then when I press 'w' all of the solid rings around my stars become white but the aura around my warpgated star stays blue even though the star is white now

**Comment by anicolao** [comment]
The problem is that in the `stars.png` file that defines all the sprites, there is only one warpgate highlight per colour. It looks like this:

![stars](https://user-images.githubusercontent.com/1145048/210659523-bb55c900-2166-4746-8aa8-c2bb1274076e.png)

with the warp gate glow on the far right of each column. So if you're blue and I change the blue highlight to white, it'll be white for all blue players :(

Maybe this indicates a problem with the approach I took, which was to recolour the sprite bitmap.

**Comment by Skilly146** [comment]
It seems like the best way to deal with this is to find the variable that stores the warpgate effect for each star (same as the star variable that stores the outline image I talked about in the original FR) and either change it via your current method or just redefine it to link to a white version provided by you. Basically it would be the exact same method I suggested in the original FR except we need to find the warp gate effect variable instead of the star outline variable. 

I don't know this variable right now but I could find it for you if you would like

**Comment by anicolao** [comment]
I do know it, and I posted its value above. It's a large sprite that contains all the ownership outlines and all the glows in one image. Each stargate sprite refers to this same bitmap with the correct coordinates for a glow from the 9th colum, ranging from x,y = 8*64, 64, up to x,y=8*64+7*64. To enable general colouring of the warpgates, either the warpgate sprite needs to be replicated 8 times at this size and leave the coordinate finding alone, or it needs to be replicated 64 times and all the coordinates computed a new way.

**Comment by anicolao** [comment]
If we wanted to do it with the 8 copies of the sprite bitmap, the relevant fix would be to patch these lines:

```
  // warp gates                                                                                                                              
  if (star.ga == 1) {                                                                                                                        
    starSprites.gate = {width: 64, height: 64, pivotX: 32, pivotY: 32, rotation: 0, scale: 1 * map.pixelRatio, image: map.starSrc, spriteX: 64*8, spriteY: col * 64 + 64};                                                                                                                        
  }
```

Instead of using `map.starSrc` which contains the bitmap above, it could be `map.starsPng[star.puid / 8]` as long as `star.puid >= 0`, and we'd have to generate 8 instances of `starsPng` with the 8 different colours according to which column we are in.

I'd happily accept a patch if this makes sense.

**Comment by anicolao** [comment]
(To be clear, when I say "patch those lines" I mean replace the `set()` method for the `starSprites.gate` property.)

**Comment by Skilly146** [comment]
Is there a reason that we wouldn't want to loop through every star and if owned by the player redefine the `starSprites.gate` variable by pointing it to one white highlight image? Thats what I meant by changing the actual variable (in this case `starSprites.gate`) that stores the final image that the website references when displaying stars. Instead of changing the source for the variable creation we can instead just change the source for player owned stars on a case by case basis.

If we use this system for the star outline as well as the warpgate highlight it would have the added benefit of full proteus support since I'm pretty sure that proteus players can have the same color/shape combo. Since we changing player stars instead of the shared source that the players stars also link to it won't effect anyone else

**Comment by anicolao** [comment]
I don't mind if you want to try it that way, and see how easy it is to get working. You could still hook the same setter and just override it differently than I'm proposing, to contain a different image and coordinates according to which group of players the highlight is in.

The code as written doesn't just handle changing the current player to white: it handles changing every player to any colour you want. It does this by recolouring _all_ the images for ownership in the 8x8 grid of the sprite, except for the star glows, which cannot be made into 8 different colours each because there is only one image. 

So the fix in your approach would be to override the setter and set a different (one of 64 possible) stargate glow according to the player's colour and the stargate glow template. It is possible to draw that into a small 8*64x64 image and set the  spriteY coordinates to 0, leaving spriteX alone. This should be done in advance to avoid generating those images over and over again, and then in the hacked assignment operator.

If you literally want a different image for every star sprite, you can't get the information you need in the location that I am talking about as only the `col` number is available (encoded in `spriteY`). You could hook sprite creation and loop over all the sprites trying to figure out which player owns each by looking at one of the other sprites, or you could rewrite the entire sprite creation function. Or, if you can find an easier way to do it, by all means put up a patch -- I do not really care how it is done, as long as we don't create new images on every draw (and ideally we'd do no new work on each draw at all, but do it all in advance somehow).

**Comment by anicolao** [comment]
> If we use this system for the star outline as well as the warpgate highlight it would have the added benefit of full proteus support since I'm pretty sure that proteus players can have the same color/shape combo. Since we changing player stars instead of the shared source that the players stars also link to it won't effect anyone else

I've not looked at the proteus source at all, but in triton the ownership ring is a single image whether it is applied to a star or a fleet. I don't see a need to change how that works, maybe youc an elaborate?

**Comment by Skilly146** [comment]
Sorry if I was unclear, my solution doesn't touch the code you mentioned. All mine would be adding is a for loop in the extension that loops through every star object in universe and checks if it is owned by the user. If the star is owned by the user than you change the `star.ownership` (Don't think its actually called that but its the variable that stores the stars outline image) and `star.gate` variable to instead be a white version of their outline and warpgate highlight using a supplied image that is an all white version of one row from the original image.

**Comment by Skilly146** [comment]
as for proteus, lets say there is a 4 player game and two of them are orange hexes. If the user, as one of the orange hexes, tried to use the 'w' hotkey it would make both the user and the other orange hex white since you're changing the actual source image. Since both orange hexes are referencing the exact same color/shape combo in the image they are both affected by the extension editing it. same color/shape combo is a bit of an edge case but since we probably have to use the same system for the warpgate highlighting we might as well avoid that problem

**Comment by anicolao** [comment]
Feel free to try to code what you're suggesting. According to my understanding of the code, there is no star outline image associated with the star. The sprite generation code generates a set of sprites that amalgamate multiple map objects into one sprite and then draw each sprite. In this way, overlapping fleets get merged, fleet text gets merged, strengths are aggregated for stars, and so on. In this way, when there is an alliance the sprite for ownership is made up of both player's sprites, one to show the owner of the star and one to show guest carriers. 

If it works as you say, try to fix it and I will happily accept a working patch.

Re: proteus it sounds like it renders completely differently. I haven't played it, is it worth trying out and trying to fix the extension to work with it? How similar is the client side, and has anyone asked Jay's permission?

**Comment by anicolao** [comment]
While I was driving I realized the `col` variable is misnamed and actually refers to the row. So the approach I was supposing would work will not work … something more like what you propose is going to be needed. Are you going to look at this, or should I take a look tonight?

**Comment by Skilly146** [comment]
Are you saying Proteus renders differently because of something you've heard elsewhere or because I said there can be multiples of the same color/shape? If it's because of what I said the only difference (that I know of) is that players get to adjust their color/shape mid game via buttons on the personal empire screen. In general I don't think Proteus's underlying code is systematically different, just the rule changes and new ui

**Comment by Skilly146** [comment]
I unfortunately still don't know JavaScript nearly well enough to be messing with the extension so it would be great if you could look at it tonight

**Comment by Skilly146** [comment]
About joining a Proteus game, I can make a custom test game when I get home. Unless you have the unlocked game creation settings from Jay I'll just have to make it for you but that's fine

**Comment by anicolao** [comment]
Sure a 1v1 game woudl be good to look at it. Unless it is very little work, I probably won't worry about proteus but I'd be more than happy if you want to fix up the extension to work with it. Are the rules better?

**Comment by anicolao** [comment]
Fixed in 516f6205a889fa4c7dbda1f9015c4b6256809420

**Comment by anicolao** [comment]
Verified in v2.0.6 on firefox macos.
